tempesta_fw/peer.h

```diff
@@ -43,29 +43,42 @@ static inline void
 tfw_peer_init(TfwPeer *p, const TfwAddr *addr)
 {
 	INIT_LIST_HEAD(&p->conn_list);
-	rwlock_init(&p->conn_lock);
+	spin_lock_init(&p->conn_lock);
 
 	memcpy(&p->addr, addr, sizeof(p->addr));
 }
 
 static inline void
 tfw_peer_add_conn(TfwPeer *p, struct list_head *conn_list)
 {
-	write_lock(&p->conn_lock);
+	spin_lock(&p->conn_lock);
 
 	list_add(conn_list, &p->conn_list);
 
-	write_unlock(&p->conn_lock);
+	spin_unlock(&p->conn_lock);
 }
 
 static inline void
 tfw_peer_del_conn(TfwPeer *p, struct list_head *conn_list)
 {
-	write_lock(&p->conn_lock);
+	spin_lock(&p->conn_lock);
 
 	list_del_init(conn_list);
 
-	write_unlock(&p->conn_lock);
+	spin_unlock(&p->conn_lock);
 }
 
+#define tfw_peer_for_each_conn(p, conn, member, cb)			\
+({									\
+	int r = 0;							\
+	spin_lock(&(p)->conn_lock);					\
+	list_for_each_entry(conn, &(p)->conn_list, member) {		\
+		r = (cb)(conn);						\
+		if (unlikely((r)))					\
+			break;						\
+	}								\
+	spin_unlock(&(p)->conn_lock);					\
+	r;								\
+})
```

Comment:
Unfortunately, we need to pass `member` to the macro and this is the only reason why I used macro instead of inline function. We don't know where exactly TfwConnection embodies list meber, so we can't get connection from a list member to pass the pointer to the callback function. Including connection.h to peer.h looks ever stronger ugliness than using the macro. Meantime, I still don't see an issue with the macro: it's small, it's does generic logic and it uses common GCC practices.